Adopt ImmersiveSpace with preferredSurroundingsEffect for 'preferredDarkness' and ImmersiveSpace for 'mrui_activeStage' due to deprecation

Phinny01 opened the pull request:
#### f97ba929583c3b390deae821fee3f3b53364103a
<pre>
Adopt ImmersiveSpace with preferredSurroundingsEffect for &apos;preferredDarkness&apos; and ImmersiveSpace for &apos;mrui_activeStage&apos; due to deprecation
<a href="https://rdar.apple.com/163514332">rdar://163514332</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=304209">https://bugs.webkit.org/show_bug.cgi?id=304209</a>

Reviewed by Richard Robinson.

Created WKFullScreenImmersiveSpaceManager to coordinate effect state between Objective-C and
SwiftUI. Added WKSurroundingsEffectView that applies preferredSurroundingsEffect on a SwiftUI view based on the
manager&apos;s current effect value. Added WKSurroundingsEffectWindow that bridges the effect view into
the UIKit fullscreen window via UIHostingController.

No ImmersiveSpace required - preferredSurroundingsEffect works in shared space for the
dark/semiDark effects WebKit needs.

* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.h: Copied from Source/WebKit/Modules/Internal/WebKitInternal.h.
* Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift: Added.
(WKSurroundingsEffectManagerWrapper.currentEffect):
(WKSurroundingsEffectManagerWrapper.surroundingsEffect):
(WKSurroundingsEffectManager.wrapper):
(WKSurroundingsEffectManager.shared):
(WKSurroundingsEffectManager.currentEffect):
(WKSurroundingsEffectWindow.effectHostingController):
(WKSurroundingsEffectWindow.setupSurroundingsEffectIfNeeded):
* Source/WebKit/Platform/spi/visionos/WKSurroundingsEffectView.swift: Added.
(WKSurroundingsEffectView.body):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenParentWindowState initWithWindow:]):
(-[WKFullScreenWindowController didEnterVideoFullscreen]):
(-[WKFullScreenWindowController didExitVideoFullscreen]):
(-[WKFullScreenWindowController _enterFullScreen:windowScene:completionHandler:]):
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
(-[WKFullScreenWindowController toggleSceneDimming]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/307258@main">https://commits.webkit.org/307258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53ad59c5553009d37ab874dda18cfc26ee748f90

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/143839 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/16512 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/8029 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/152507 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/97078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/16997 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/16408 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/110609 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/97078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/146802 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/13044 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/129245 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/91527 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/12529 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/10238 "Passed tests") | [✅ 🛠 wpe-libwebrtc](https://ews-build.webkit.org/#/builders/172/builds/2509 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/121971 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/154819 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/16368 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/6907 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/118618 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/16403 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/13761 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/118973 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/30498 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/14894 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/127073 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/71781 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/15989 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/5562 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/15723 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/79760 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/15935 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/15788 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -68,6 +68,16 @@
 #import "MRUIKitSPI.h"
+
+// Forward declarations for ImmersiveSpace API (visionOS 2.0+)
+// TODO: Remove once these are in MRUIKitSPI.h
```
that TODO comment is unlikely to be correct.
Those interfaces aren't SPI

We aren't building for visionOS < 2.0 anymore nor will we ever be going to. Please use the proper include for it, no need for forward declarations

Phinny01 replied:
You're right. They are SwiftUI APIs. I have corrected it now.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
```diff
@@ -52,3 +52,5 @@ SOFT_LINK_CLASS_FOR_HEADER(WebKit, WKIdentityDocumentPresentmentMobileDocumentRe
 SOFT_LINK_CLASS_FOR_HEADER(WebKit, WKIdentityDocumentPresentmentRequestAuthenticationCertificate);
 SOFT_LINK_CLASS_FOR_HEADER(WebKit, WKTextAnimationManager)
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, WKFullScreenImmersiveSpaceManager);
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, WKSurroundingsEffectWindow);
```
I don't think your classes need to be in `WebKitSwift`. They can be in WebKit itself. You should be able to follow `WKMaterialHostingSupport` as an example.

Phinny01 replied:
Thanks that was helpful.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenImmersiveSpaceDelegate.h:
```diff
@@ -0,0 +1,47 @@
+@interface WKFullScreenImmersiveSpaceManager : NSObject
++ (instancetype)shared;
+@property (nonatomic) NSInteger currentEffect;
+- (void)setEffect:(NSInteger)effect;
```
This should be a single readwrite property.

Phinny01 replied:
I addressed this

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -121,6 +125,19 @@ static bool useSpatialFullScreenTransition()
+{
+    WKFullScreenImmersiveSpaceManager *manager = [getWKFullScreenImmersiveSpaceManagerClassSingleton() shared];
+    NSInteger effect = manager.currentEffect;
+    return (MRUIDarknessPreference)effect;
```
What's the reason for the casting? Why not use the actual type in the Swift interface?

Phinny01 replied:
You're right we don't need to

pxlcoder commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -1835,6 +1834,8 @@
 		96EE6CE82DFA4980000F90BD /* WebExtensionAPIBookmarks.h in Headers */ = {isa = PBXBuildFile; fileRef = 96EE6CE72DFA4975000F90BD /* WebExtensionAPIBookmarks.h */; };
 		9732D19C2ECFD0B600AC856D /* _WKCaptionStyleMenuControllerAVKitMac.mm in Sources */ = {isa = PBXBuildFile; fileRef = 97AD28872ECFA49800039097 /* _WKCaptionStyleMenuControllerAVKitMac.mm */; };
+		976544222EF2847D00BD8947 /* WKFullScreenImmersiveSpaceDelegate.swift in Sources */ = {isa = PBXBuildFile; fileRef = 9765441E2EF22B0B00BD8947 /* WKFullScreenImmersiveSpaceDelegate.swift */; };
+		976544262EF3514300BD8947 /* WKFullScreenImmersiveSpaceDelegate.h in Headers */ = {isa = PBXBuildFile; fileRef = 976544232EF34BDE00BD8947 /* WKFullScreenImmersiveSpaceDelegate.h */; settings = {ATTRIBUTES = (Private, ); }; };
```
We certainly don't want this to be a private header because that means people outside WebKit could see and use it.

(following the pattern of `WKMaterialHostingSupport` mentioned above should address this)

Phinny01 replied:
Right! thanks!

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+//INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+//CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+//ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+//THE POSSIBILITY OF SUCH DAMAGE.
```
There should be a single space after `//` on each line.

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+//CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+//ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+//THE POSSIBILITY OF SUCH DAMAGE.
+//
```
No need for the trailing `//`

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+/// Manages surroundings effect state for fullscreen web content in visionOS.
+///
+/// This singleton coordinates the surroundings dimming effect between Objective-C/UIKit
+/// and SwiftUI layers, replacing the deprecated MRUIStage.preferredDarkness API.
```
I don't think this comment is really necessary. Though if you decide to keep it, I would certainly remove the "replacing the deprecated MRUIStage.preferredDarkness API" – that detail is not relevant over time.

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+    /// The current surroundings effect level (0=none, 1=semiDark, 2=dark, 3=ultraDark).
+    @Published
+    @objc
+    public dynamic var currentEffect: Int = 0
```
This should be an actual enumeration rather than an `Int`. Then the comment is also unnecessary.

Again `WKMaterialHostingSupport` could serve as inspiration (`WKHostedMaterialEffectType`).

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+            return .dark
+        case 3:
+            return .ultraDark
+        default:
```
Once we define a real enumeration, this can be `@unknown default`.

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+        super.init()
+    }
+
+    var swiftUIEffect: SurroundingsEffect? {
```
We're in SwiftUI code, and this property isn't exposed outside of SwiftUI code, so SwiftUI in the name doesn't add value. `fileprivate` because this variable is only in used in this file.

```suggestion
    fileprivate var surroundingsEffect: SurroundingsEffect? {
```

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+
+    private override init() {
+        super.init()
+    }
```
Is the idea behind this to hide the initializer so clients must use `shared`? Otherwise it seems unnecessary.

Phinny01 replied:
Yes exactly

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+
+struct WKSurroundingsEffectView: View {
+    @ObservedObject
+    var manager = WKFullScreenImmersiveSpaceManager.shared
```
Is the variable necessary? If so, I think it should be `private`.

Phinny01 replied:
Yes I think it is necessary-  without it, the UI would be stale but you are right it should be private

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+
+    required init?(coder: NSCoder) {
+        super.init(coder: coder)
+    }
```
Are these initializers necessary? They don't do anything other than the original behavior.

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+/// A UIWindow subclass that applies SwiftUI surroundings effects to fullscreen web content.
+///
+/// This window bridges SwiftUI's preferredSurroundingsEffect modifier into the UIKit-based
+/// fullscreen window hierarchy, enabling surroundings dimming without using deprecated APIs.
```
Also not sure this comment is necessary. 

"A UIWindow subclass that applies SwiftUI surroundings effects to fullscreen web content", is inaccurate since this class is used in other parts of the code to apply "SwiftUI surroundings effects to fullscreen web content". This code alone does not know about fullscreen.

And same comment above about mentioning the deprecated APIs.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenImmersiveSpaceDelegate.h:
```diff
@@ -0,0 +1,46 @@
+#import <Foundation/Foundation.h>
+#import <TargetConditionals.h>
+
+#if TARGET_OS_VISION
```
We should use our own `PLATFORM` macro and remove the import of `TargetConditionals.h`.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -121,6 +125,21 @@ static bool useSpatialFullScreenTransition()
+static constexpr NSInteger SurroundingsEffectLevelNone = 0;
+static constexpr NSInteger SurroundingsEffectLevelSemiDark = 1;
+static constexpr NSInteger SurroundingsEffectLevelDark = 2;
+static constexpr NSInteger SurroundingsEffectLevelUltraDark = 3;
```
These should be an actual enum as mentioned in an earlier comment.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -121,6 +125,21 @@ static bool useSpatialFullScreenTransition()
+static void setSurroundingsEffect(NSInteger effect)
+{
+    [WKFullScreenImmersiveSpaceManager shared].currentEffect = effect;
+}
```
I don't think we need these methods. The callers could just inline the single line of code?

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+@MainActor
+@objc(WKFullScreenImmersiveSpaceManager)
+@available(visionOS 1.0, *)
+public class WKFullScreenImmersiveSpaceManager: NSObject, ObservableObject {
```
I think this class could be renamed to remove the mention of "FullScreen" since its implementation is independent of fullscreen.

pxlcoder commented on Source/WebKit/Platform/spi/visionos/WKFullScreenImmersiveSpaceDelegate.swift:
```diff
@@ -0,0 +1,127 @@
+@MainActor
+@objc(WKFullScreenImmersiveSpaceManager)
+@available(visionOS 1.0, *)
+public class WKFullScreenImmersiveSpaceManager: NSObject, ObservableObject {
```
On further thought is this class truly necessary? Could the current effect simply live on the `WKSurroundingsEffectWindow`?

Phinny01 replied:
Yes I think it is necessary. If we stored it directly on the window, we'd need manual KVO or notification mechanisms to make SwiftUI react to Objective-C property changes which I think would be more complex than this.

pxlcoder reviewed (commented):
It's unclear why the new file is called `WKFullScreenImmersiveSpace**Delegate**`. There's no use of the delegate pattern here.

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.h:
```diff
@@ -0,0 +1,55 @@
+    WKSurroundingsEffectTypeUltraDark = 3,
+};
+
+NS_ASSUME_NONNULL_BEGIN
```
this should be 

```
NS_HEADER_AUDIT_BEGIN(nullability, sendability)
```

instead (and likewise for the END one)

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.h:
```diff
@@ -0,0 +1,55 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface WKSurroundingsEffectManager : NSObject
```
this type itself should just be NS_SWIFT_UI_ACTOR, and then the methods/properties don’t have to explicitly say so

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+
+#if os(visionOS)
+
+import Combine
```
There is no need to use Combine; The Observation framework should be preferred instead

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+    case semiDark = 1
+    case dark = 2
+    case ultraDark = 3
+}
```
this is a duplicate definition of the other WKSurroundingsEffectType type and should be deleted

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+@MainActor
+@objc(WKSurroundingsEffectManager)
+@available(visionOS 1.0, *)
+class WKSurroundingsEffectManager: NSObject, ObservableObject {
```
this needs to just be `@objc @implementation extension WKSurroundingsEffectManager`

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+@objc(WKSurroundingsEffectManager)
+@available(visionOS 1.0, *)
+class WKSurroundingsEffectManager: NSObject, ObservableObject {
+    @objc
```
and then no `@objc` on any of the members are needed

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+
+    @Published
+    @objc
+    dynamic var currentEffect: WKSurroundingsEffectType = .none
```
@Published and `dynamic` can be removed once Observation is used instead of Combine

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+    @objc
+    dynamic var currentEffect: WKSurroundingsEffectType = .none
+
+    private override init() {
```
this is unnecessary; the init should just be marked as NS_UNAVAILABLE on the objc side

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    fileprivate var surroundingsEffect: SurroundingsEffect? {
+        switch self.currentEffect {
```
no `self.`

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+    fileprivate var surroundingsEffect: SurroundingsEffect? {
+        switch self.currentEffect {
+        case .none:
+            return nil
```
no need for any `return` keywords

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+    }
+}
+
+struct WKSurroundingsEffectView: View {
```
this should be in its own file

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+
+struct WKSurroundingsEffectView: View {
+    @ObservedObject
+    private var manager = WKSurroundingsEffectManager.shared
```
this should be injected from the environment

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+    private var manager = WKSurroundingsEffectManager.shared
+
+    var body: some View {
+        Color.clear
```
why not just `EmptyView()`?

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+
+@objc(WKSurroundingsEffectWindow)
+@available(visionOS 1.0, *)
+class WKSurroundingsEffectWindow: UIWindow {
```
ditto about @objc @implementation

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    private func setupSurroundingsEffect() {
+        guard let rootVC = rootViewController else {
```
```suggestion
        guard let rootViewController else {
```

rr-codes reviewed (changes requested):
(requesting changes per my above comments)

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.h:
```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#import <Foundation/Foundation.h>
+#import <wtf/PlatformHave.h>
```
```suggestion
#import <wtf/Platform.h>
```

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,87 @@
+
+#if os(visionOS)
+
+import Observation
```
```suggestion
internal import Observation
```

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,87 @@
+
+import Observation
+@_weakLinked internal import SwiftUI
+import UIKit
```
```suggestion
internal import UIKit
```

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,87 @@
+
+    class func shared() -> Self {
+        shared as? Self ?? Self()
+    }
```
```suggestion
    class func shared() -> WKSurroundingsEffectManager {
        shared
    }
```

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,87 @@
+}
+
+@available(visionOS 1.0, *)
+extension WKSurroundingsEffectManager: Observable {
```
this is wrong; an Objective-C type should not be made Observable, nor should Observable be directly used as conformance

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,87 @@
+
+    var currentEffect: WKSurroundingsEffectType = .none
+
+    final var surroundingsEffect: SurroundingsEffect? {
```
```suggestion
   @nonobjc
   var surroundingsEffect: SurroundingsEffect? {
```

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,87 @@
+    private final var effectHostingController: UIHostingController<WKSurroundingsEffectView>?
+
+    private func setupSurroundingsEffectIfNeeded() {
+        guard effectHostingController == nil, let rootViewController = rootViewController else {
```
```suggestion
        guard effectHostingController == nil, let rootViewController else {
```

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,110 @@
+
+private let sharedWrapper = WKSurroundingsEffectManagerWrapper()
+
+extension WKSurroundingsEffectManager {
```
why is this wrapper needed?

Phinny01 replied:
The wrapper provides @Observable support for the Objective-C WKSurroundingsEffectManager class. The @Observable macro only works with Swift classes, so the wrapper acts as an observable bridge, enabling reactive SwiftUI updates via .environment(wrapper).

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,110 @@
+        set { sharedWrapper.currentEffect = newValue }
+    }
+
+    override init() {}
```
why is this needed?

Phinny01 replied:
This is required to override the Objective-C class's designated initializer, allowing us to create shared​Instance. The @implementation extension needs this to satisfy the class's initialization contract.

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,110 @@
+@implementation
+extension WKSurroundingsEffectWindow {
+    private final var effectHostingController: UIHostingController<WKSurroundingsEffectView>?
+    private final var effectManagerWrapper: WKSurroundingsEffectManagerWrapper?
```
why is this needed when it’s always just the singleton instance?

Phinny01 replied:
You're right. This is unnecessary.

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,110 @@
+        let effectView = WKSurroundingsEffectView()
+            .environment(wrapper)
+        let hostingController = UIHostingController(rootView: effectView)
+        hostingController.view.backgroundColor = UIColor.clear
```
```suggestion
        hostingController.view.backgroundColor = .clear
```

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,110 @@
+        hostingController.view.backgroundColor = UIColor.clear
+        hostingController.view.isUserInteractionEnabled = false
+        hostingController.view.frame = rootViewController.view.bounds
+        hostingController.view.autoresizingMask = [UIView.AutoresizingMask.flexibleWidth, UIView.AutoresizingMask.flexibleHeight]
```
```suggestion
        hostingController.view.autoresizingMask = [.flexibleWidth, .flexibleHeight]
```

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift:
```diff
@@ -0,0 +1,110 @@
+        hostingController.view.frame = rootViewController.view.bounds
+        hostingController.view.autoresizingMask = [UIView.AutoresizingMask.flexibleWidth, UIView.AutoresizingMask.flexibleHeight]
+
+        effectHostingController = hostingController as? UIHostingController<WKSurroundingsEffectView>
```
why is this cast needed?

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffectView.swift:
```diff
@@ -0,0 +1,40 @@
+
+    var body: some View {
+        let effect = wrapper.surroundingsEffect
+
+        return EmptyView()
+            .preferredSurroundingsEffect(effect)
+    }
```
```suggestion
    var body: some View {
        EmptyView()
            .preferredSurroundingsEffect(wrapper.surroundingsEffect)
    }
```

rr-codes commented on Source/WebKit/Platform/spi/visionos/WKSurroundingsEffectView.swift:
```diff
@@ -0,0 +1,40 @@
+    var body: some View {
+        let effect = wrapper.surroundingsEffect
+
+        return EmptyView()
```
why is this an empty view?

Phinny01 replied:
We only need a SwiftUI view to apply the .preferred​Surroundings​Effect() modifier. Empty​View is used because the view itself doesn't render any content - it exists solely to propagate the surroundings effect preference through the view hierarchy to the window.

Commit:
Adopt ImmersiveSpace with preferredSurroundingsEffect for 'preferredDarkness' and ImmersiveSpace for 'mrui_activeStage' due to deprecation
rdar://163514332
https://bugs.webkit.org/show_bug.cgi?id=304209

Reviewed by Richard Robinson.

Created WKFullScreenImmersiveSpaceManager to coordinate effect state between Objective-C and
SwiftUI. Added WKSurroundingsEffectView that applies preferredSurroundingsEffect on a SwiftUI view based on the
manager's current effect value. Added WKSurroundingsEffectWindow that bridges the effect view into
the UIKit fullscreen window via UIHostingController.

No ImmersiveSpace required - preferredSurroundingsEffect works in shared space for the
dark/semiDark effects WebKit needs.

Mark Swift-only stored properties as final in WKSeparatedImageView - Using override in
@objc @implementation extensions triggers stricter validation of all such extensions, requiring
Swift-only properties to be marked final.

* Source/WebKit/Modules/Internal/WebKitInternal.h:
* Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.h: Copied from Source/WebKit/Modules/Internal/WebKitInternal.h.
* Source/WebKit/Platform/spi/visionos/WKSurroundingsEffect.swift: Added.
(WKSurroundingsEffectManagerWrapper.currentEffect):
(WKSurroundingsEffectManagerWrapper.surroundingsEffect):
(WKSurroundingsEffectManager.wrapper):
(WKSurroundingsEffectManager.shared):
(WKSurroundingsEffectManager.currentEffect):
(WKSurroundingsEffectWindow.effectHostingController):
(WKSurroundingsEffectWindow.setupSurroundingsEffectIfNeeded):
* Source/WebKit/Platform/spi/visionos/WKSurroundingsEffectView.swift: Added.
(WKSurroundingsEffectView.body):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(-[WKFullScreenParentWindowState initWithWindow:]):
(-[WKFullScreenWindowController didEnterVideoFullscreen]):
(-[WKFullScreenWindowController didExitVideoFullscreen]):
(-[WKFullScreenWindowController _enterFullScreen:windowScene:completionHandler:]):
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
(-[WKFullScreenWindowController toggleSceneDimming]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/307258@main
